### ISSUE TITLE
[v6][Bug]: Different behavior using location state and reload

### ISSUE THREAD

**Issue by rafaelgorr** [remix-run/react-router#8575]
### What version of React Router are you using?

6.2.1

### Steps to Reproduce

With 6.2.1:
https://stackblitz.com/edit/react-ts-ekdemx?file=index.tsx

With 5.2.0:
https://stackblitz.com/edit/react-ts-5gy6ny?file=index.tsx

Click on button HERE and then, in the /hello route, reload stackblitz browser.

### Expected Behavior

Both v5 and v6 should show "Not found" message after reload in /hello route. 

### Actual Behavior

After you click on reload button, v5 shows "Not found" and v6 shows "Hello Some Name!".

Basically after reload, v6 keeps the state value _'name'_ and v5 resets it as undefined. I tested locally using Brave browser and it also happen when you go back e go forward using v6, v5 stills returning undefined, but for some reason on stackblitz environment when you go forward work properly using v6.

**Comment by timdorr** [comment]
This is because you're using HashRouter. In v6, it uses history@5, which adds support for `location.state`. This is intentional behavior for other router types, which is now the same when using HashRouter.
